### ISSUE TITLE
Live capture: Resource temporarily unavailable

### ISSUE THREAD

**Issue by dbrgn** [protectwise/pcap-async#21]
I can parse pcap files just fine (using `Handle::file_capture("/tmp/ble.pcap")`), but when I try to capture from a live device:

    let handle = Handle::live_capture("bluetooth0").expect("No handle created");

...I get the following error, even when running with root permissions:

    Err(LibPcapError("Can\'t receive packet: Resource temporarily unavailable"))

At the same time, capturing this device with Wireshark works fine.

Do you have an idea what the reason could be, or how to debug this?

**Comment by dbcfd** [comment]
Is this on arm/raspberry pi?

**Comment by dbrgn** [comment]
No, on Arch Linux with a BLE dongle over USB.

**Comment by dbrgn** [comment]
From the log:

```
[2020-06-06T13:00:02Z INFO  pcap_async::handle] Live stream created for interface bluetooth0
[2020-06-06T13:00:02Z DEBUG pcap_async::packet::future] Dispatch returned with -1
[2020-06-06T13:00:02Z ERROR pcap_async::pcap_util] LibPcap encountered an error: Can't receive packet: Resource temporarily unavailable
[2020-06-06T13:00:02Z ERROR pcap_async::packet::future] Error encountered when calling pcap_dispatch: Libpcap encountered an error: Can't receive packet: Resource temporarily unavailable
```

This doesn't seem to happen with an ethernet interface.

Note that `bluetooth0` is one of the interfaces returned by `pcap_async::Info::all()`, so it should be valid.

**Comment by dbrgn** [comment]
This C program seems to capture something:

```c
#include <stdio.h>
#include <pcap.h>

int main(int argc, char *argv[]) {
    char *dev = "bluetooth0\0";
    char errbuf[PCAP_ERRBUF_SIZE];

    // Open device
    printf("Opening device: %s\n", dev);
    pcap_t *handle = pcap_open_live(dev, BUFSIZ, 1, 1000, errbuf);
    if (handle == NULL) {
        fprintf(stderr, "Couldn't open device %s: %s\n", dev, errbuf);
        return 2;
    }

    // Capture some packets
    struct pcap_pkthdr header;
    const u_char *packet;
    for (int i = 0; i < 10; i++) {
        packet = pcap_next(handle, &header);
        printf("%d: Captured a packet with length of [%d]\n", i, header.len);
    }

    pcap_close(handle);
    return 0;
}
```

Output:

```
danilo@x2000:tmp$ clang capture.c -lpcap -o capture
danilo@x2000:tmp$ ./capture
Opening device: bluetooth0
0: Captured a packet with length of [8]
1: Captured a packet with length of [11]
2: Captured a packet with length of [11]
3: Captured a packet with length of [11]
4: Captured a packet with length of [11]
5: Captured a packet with length of [11]
6: Captured a packet with length of [262]
7: Captured a packet with length of [262]
8: Captured a packet with length of [262]
9: Captured a packet with length of [8]
```

**Comment by dbcfd** [comment]
pcap-async is using the file descriptor to determine if the device has packets ready. Will try to figure out why that file descriptor isn't working properly. I also need to adjust the fd check to be < 0.

**Comment by dbcfd** [comment]
Bluetooth sockets might also not support non block, so will have to add nonblock back to config and use it here https://github.com/protectwise/pcap-async/blob/master/src/stream.rs#L36

**Comment by dbrgn** [comment]
Ah, of course, async is a bit more complex than just waiting for `pcap_next` :slightly_smiling_face:

**Comment by dbrgn** [comment]
Note also that I'm using current master (with Smol), not the released version. Not sure if that matters...

**Comment by dbrgn** [comment]
I played around with the code a bit. It works fine if I don't put the socket (handle) in non-blocking mode. However, in that case the async executor would be blocked when polling the socket, right? (I'm not very familiar with `mio`, so I don't know how this works in detail.) If yes, would the polling need to be moved to its own thread, or are there some tricks to avoid that?

(Probably unrelated: I noticed there's [a newer version of mio](https://tokio.rs/blog/2019-12-mio-v0.7-alpha.1/) with some changes with regard to socket events...)
